sound/soc/sof/virtio-be.c

```diff
@@ -5,122 +5,978 @@
+		return;
+	}
+	vq = &vbe->vqs[vq_idx];
+	entry = list_first_entry(&vbe->posn_list,
+				 struct vbs_sof_posn, list);
+
+	ret = sbe_fill_posn_vqbuf(vbe, vq, &entry->pos, &endchain);
+	spin_unlock_irqrestore(&vbe->posn_lock, flags);
+	if (endchain)
+		virtio_vq_endchains(vq, 1);
+}
+
+/* validate component IPC */
+static int sbe_ipc_comp(struct snd_sof_dev *sdev, int vm_id,
+			struct sof_ipc_hdr *hdr)
+{
+	/*TODO validate host comp id range based on vm_id */
+
+	/* Nothing to be done */
 	return 0;
 }
 
-static void sof_virtio_remove(struct virtio_device *dev)
+/*
+ * This function is to get the BE dai link for the GOS
+ * It uses the dai_link name to find the BE dai link.
+ * The Current dai_link name "vm_dai_link" is for the GOS,
+ * which means only one Virtual Machine is supported.
+ * And the VM only support one playback pcm and one capture pcm.
+ * After we switch to the new topology, we can support multiple
+ * VMs and multiple PCM streams for each VM.
+ * This function may be abandoned after switching to the new topology.
+ */
+static struct snd_pcm_substream *
+sbe_get_substream(struct snd_sof_dev *sdev,
+		  struct snd_soc_pcm_runtime *rtd, int direction)
+{
+	struct snd_pcm *pcm;
+	struct snd_pcm_str *stream;
+	struct snd_pcm_substream *substream = NULL;
+	struct snd_soc_dai_link *dai_link;
+	struct snd_soc_card *card = sdev->card;
+
+	list_for_each_entry(rtd, &card->rtd_list, list) {
+		pcm = rtd->pcm;
+		if (!pcm)
+			continue;
+
+		stream = &pcm->streams[direction];
+		if (!stream)
+			continue;
+
+		substream = stream->substream;
+		if (substream) {
+			dai_link = rtd->dai_link;
+			if (strcmp(dai_link->name, "vm_dai_link") == 0)
+				return substream;
+		}
+	}
+
+	return NULL;
+}
+
+static int sbe_pcm_open(struct snd_sof_dev *sdev,
+			void *ipc_data, int vm_id)
 {
-	/* remove topology from fe via virtqueue */
+	/*
+	 * TO re-use the sof callback for pcm, we should find a proper
+	 * substream and do the correct setting for the substream.
+	 * As there is no FE link substream in SOS for GOS (GOS FE link
+	 * substreams are created in GOS and SOS will never see it), let's
+	 * use BE link substream in SOS for the callbacks.
+	 * This is save because the BE link substream is created dedicated for
+	 * GOS in machine driver.
+	 */
+	struct snd_pcm_substream *substream;
+	struct snd_pcm_runtime *runtime;
+	struct snd_sof_pcm *spcm;
+	struct snd_soc_pcm_runtime *rtd = NULL;
+	u32 comp_id;
+	size_t size;
+	int direction;
+
+	comp_id = get_comp_id(sdev, ipc_data);
+
+	spcm = snd_sof_find_spcm_comp(sdev, comp_id, &direction);
+	if (!spcm)
+		return -ENODEV;
+	mutex_lock(&spcm->mutex);
+	substream = sbe_get_substream(sdev, rtd, direction);
+	if (!substream || !rtd)
+		return -ENODEV;
+	if (substream->ref_count > 0)
+		return -EBUSY;
+	substream->ref_count++;	/* set it used */
+	runtime = kzalloc(sizeof(*runtime), GFP_KERNEL);
+	if (!runtime)
+		return -ENOMEM;
+	size = PAGE_ALIGN(sizeof(struct snd_pcm_mmap_status));
```

Comment:
Also this function needs split into function blocks separated by newlines and with comments at each block